File: Lab_2/timeconsuming.R
library(parallel)
library(microbenchmark)

time_consuming_fun <- function() {
  pause_time <- runif(1, 1, 5)  # Random pause between 1 and 5 seconds
  Sys.sleep(pause_time)
}

run_experiment <- function(n) {
  serial_times <- microbenchmark(
    for (i in 1:n) time_consuming_fun(),
    times = 1
  )

  parallel_times <- microbenchmark(
    mclapply(1:n, function(i) time_consuming_fun(), mc.cores = detectCores()),
    times = 1
  )

  serial_time <- summary(serial_times)[, "median"]
  parallel_time <- summary(parallel_times)[, "median"]
  
  speedup <- serial_time / parallel_time
  efficiency <- speedup / detectCores()

  cat("Number of iterations:", n, "\n")
  cat("Serial time:", serial_time, "seconds\n")
  cat("Parallel time:", parallel_time, "seconds\n")
  cat("Speedup:", speedup, "\n")
  cat("Efficiency:", efficiency, "\n\n")
}

run_experiment(100)
run_experiment(1000)
run_experiment(10000)

File: Lab_2/timeconsuming.py
import time
import multiprocessing

def time_consuming_fun(duration):
    time.sleep(duration)

def run_experiment(n):
    start_time = time.time()
    for _ in range(n):
        time_consuming_fun(1 + 4 * (time.time() % 5))
    serial_time = time.time() - start_time

    start_time = time.time()
    with multiprocessing.Pool() as pool:
        pool.map(time_consuming_fun, [1 + 4 * (time.time() % 5) for _ in range(n)])
    parallel_time = time.time() - start_time

    speedup = serial_time / parallel_time
    efficiency = speedup / multiprocessing.cpu_count()
    return serial_time, parallel_time, speedup, efficiency

n_small = 10
n_medium = 100
n_large = 1000
print("Small")
serial_time_small, parallel_time_small, speedup_small, efficiency_small = run_experiment(n_small)
serial_time_medium, parallel_time_medium, speedup_medium, efficiency_medium = run_experiment(n_medium)
serial_time_large, parallel_time_large, speedup_large, efficiency_large = run_experiment(n_large)

print(f"Results for n = {n_small}:")
print(f"Serial Time: {serial_time_small} seconds")
print(f"Parallel Time: {parallel_time_small} seconds")
print(f"Speedup: {speedup_small}")
print(f"Efficiency: {efficiency_small}\n")

print(f"Results for n = {n_medium}:")
print(f"Serial Time: {serial_time_medium} seconds")
print(f"Parallel Time: {parallel_time_medium} seconds")
print(f"Speedup: {speedup_medium}")
print(f"Efficiency: {efficiency_medium}\n")

print(f"Results for n = {n_large}:")
print(f"Serial Time: {serial_time_large} seconds")
print(f"Parallel Time: {parallel_time_large} seconds")
print(f"Speedup: {speedup_large}")
print(f"Efficiency: {efficiency_large}")
